src/zenml/cli/base.py

```diff
@@ -629,12 +629,152 @@ def migrate_database(skip_default_registrations: bool = False) -> None:
     )
-    if store_config.type == StoreType.SQL:
-        BaseZenStore.create_store(
-            store_config, skip_default_registrations=skip_default_registrations
-        )
-        cli_utils.declare("Database migration finished.")
-    else:
+
+    # check if migration is needed by comparing the DB head with current
+    # using alembic current and alembic heads
+    # if the DB head is not the same as the current, then migration is needed
+    # if the DB head is the same as the current, then no migration is needed
+    from alembic.migration import MigrationContext
+    from sqlalchemy import create_engine
+
+    if store_config.type == StoreType.REST:
         cli_utils.warning(
             "Unable to migrate database while connected to a ZenML server."
         )
+        return
+
+    # replace the start of the url from mysql:// to mysql+pymysql://
+    url = store_config.url
+    if url.startswith("mysql://"):
+        url = "mysql+pymysql://" + url[8:]
+
+    try:
+        engine = create_engine(url)
+        conn = engine.connect()
+        context = MigrationContext.configure(conn)
+        current_rev = context.get_current_revision()
+        head = context.get_current_heads()[0]
+
+        logger.debug(f"Current revision: {current_rev}")
+        logger.debug(f"Head revision: {head}")
+
+        if current_rev != head:
+            logger.debug("Database migration needed. Performing a DB dump.")
+            mysql_dump_to_s3(store_config.url)
+    except Exception as e:
+        # the first time a deployment is done and the DB doesn't exist yet.
+        logger.debug(
+            f"Database doesn't exist yet. Skipping backup.\n Error: {e}"
+        )
+
+    try:
+        BaseZenStore.create_store(
+            store_config, skip_default_registrations=skip_default_registrations
+        )
+        logger.debug("Database migration successful.")
+    except Exception as e:
+        mysql_restore_from_s3(store_config.url)
+        cli_utils.error(f"Database migration failed with error: {e}")
+
+
+def mysql_dump_to_s3(url: str) -> None:
+    """Dump a mysql database into an s3 bucket directory.
+
+    Args:
+        url: The url of the database to dump
+    """
+    import pymysql
+
+    # get the database password, host without port and username from the URL
+    # URL of form mysql://<username>:<password>@<host>:<port>/<database>
+    db_password = url.split(":")[2].split("@")[0]
+    db_host = url.split("@")[1].split(":")[0]
+    db_name = url.split("/")[3]
+    db_user = url.split(":")[1].split("//")[1]
+
+    connection = pymysql.connect(
+        host=db_host,
+        user=db_user,
+        password=db_password,
+    )
+
+    # check if the database exists
+    # if it does not exist, return
+    with connection.cursor() as cursor:
+        result = cursor.execute(f'SHOW DATABASES LIKE "{db_name}"')
+        if result == 0:
+            logger.debug("Database does not exist. No backup created.")
+            return
```

Comment:
same thing here: you can reuse a lot of the existing code in the `SqlZenStore` to achieve this instead of reinventing the wheel and duplicating the code here.

Reply:
Makes the code cleaner, I like it. Thanks!